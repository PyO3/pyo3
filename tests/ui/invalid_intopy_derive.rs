use pyo3::{IntoPyObject, IntoPyObjectRef};

#[derive(IntoPyObject, IntoPyObjectRef)]
struct Foo();

#[derive(IntoPyObject, IntoPyObjectRef)]
struct Foo2 {}

#[derive(IntoPyObject, IntoPyObjectRef)]
enum EmptyEnum {}

#[derive(IntoPyObject, IntoPyObjectRef)]
enum EnumWithEmptyTupleVar {
    EmptyTuple(),
    Valid(String),
}

#[derive(IntoPyObject, IntoPyObjectRef)]
enum EnumWithEmptyStructVar {
    EmptyStruct {},
    Valid(String),
}

#[derive(IntoPyObject, IntoPyObjectRef)]
#[pyo3(transparent)]
struct EmptyTransparentTup();

#[derive(IntoPyObject, IntoPyObjectRef)]
#[pyo3(transparent)]
struct EmptyTransparentStruct {}

#[derive(IntoPyObject, IntoPyObjectRef)]
enum EnumWithTransparentEmptyTupleVar {
    #[pyo3(transparent)]
    EmptyTuple(),
    Valid(String),
}

#[derive(IntoPyObject, IntoPyObjectRef)]
enum EnumWithTransparentEmptyStructVar {
    #[pyo3(transparent)]
    EmptyStruct {},
    Valid(String),
}

#[derive(IntoPyObject, IntoPyObjectRef)]
#[pyo3(transparent)]
struct TransparentTupTooManyFields(String, String);

#[derive(IntoPyObject, IntoPyObjectRef)]
#[pyo3(transparent)]
struct TransparentStructTooManyFields {
    foo: String,
    bar: String,
}

#[derive(IntoPyObject, IntoPyObjectRef)]
enum EnumWithTransparentTupleTooMany {
    #[pyo3(transparent)]
    EmptyTuple(String, String),
    Valid(String),
}

#[derive(IntoPyObject, IntoPyObjectRef)]
enum EnumWithTransparentStructTooMany {
    #[pyo3(transparent)]
    EmptyStruct {
        foo: String,
        bar: String,
    },
    Valid(String),
}

#[derive(IntoPyObject, IntoPyObjectRef)]
#[pyo3(unknown = "should not work")]
struct UnknownContainerAttr {
    a: String,
}

#[derive(IntoPyObject, IntoPyObjectRef)]
union Union {
    a: usize,
}

#[derive(IntoPyObject, IntoPyObjectRef)]
enum UnitEnum {
    Unit,
}

#[derive(IntoPyObject, IntoPyObjectRef)]
struct TupleAttribute(#[pyo3(attribute)] String, usize);

#[derive(IntoPyObject, IntoPyObjectRef)]
struct TupleItem(#[pyo3(item)] String, usize);

#[derive(IntoPyObject, IntoPyObjectRef)]
struct StructAttribute {
    #[pyo3(attribute)]
    foo: String,
}

#[derive(IntoPyObject, IntoPyObjectRef)]
#[pyo3(transparent)]
struct StructTransparentItem {
    #[pyo3(item)]
    foo: String,
}

#[derive(IntoPyObject)]
#[pyo3(transparent)]
struct StructTransparentIntoPyWith {
    #[pyo3(into_py_with = into)]
    foo: String,
}

#[derive(IntoPyObjectRef)]
#[pyo3(transparent)]
struct StructTransparentIntoPyWithRef {
    #[pyo3(into_py_with = into_ref)]
    foo: String,
}

#[derive(IntoPyObject)]
#[pyo3(transparent)]
struct TupleTransparentIntoPyWith(#[pyo3(into_py_with = into)] String);

#[derive(IntoPyObject)]
enum EnumTupleIntoPyWith {
    TransparentTuple(#[pyo3(into_py_with = into)] usize),
}

#[derive(IntoPyObject)]
enum EnumStructIntoPyWith {
    #[pyo3(transparent)]
    TransparentStruct {
        #[pyo3(into_py_with = into)]
        a: usize,
    },
}

fn main() {}
