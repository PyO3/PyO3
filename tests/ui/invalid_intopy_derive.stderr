error: cannot derive `IntoPyObject` for empty structs
 --> tests/ui/invalid_intopy_derive.rs:4:11
  |
4 | struct Foo();
  |           ^^

error: cannot derive `IntoPyObject` for empty structs
 --> tests/ui/invalid_intopy_derive.rs:7:13
  |
7 | struct Foo2 {}
  |             ^^

error: cannot derive `IntoPyObject` for empty enum
  --> tests/ui/invalid_intopy_derive.rs:10:6
   |
10 | enum EmptyEnum {}
   |      ^^^^^^^^^

error: cannot derive `IntoPyObject` for empty variants
  --> tests/ui/invalid_intopy_derive.rs:14:5
   |
14 |     EmptyTuple(),
   |     ^^^^^^^^^^

error: cannot derive `IntoPyObject` for empty variants
  --> tests/ui/invalid_intopy_derive.rs:20:5
   |
20 |     EmptyStruct {},
   |     ^^^^^^^^^^^

error: cannot derive `IntoPyObject` for empty structs
  --> tests/ui/invalid_intopy_derive.rs:26:27
   |
26 | struct EmptyTransparentTup();
   |                           ^^

error: cannot derive `IntoPyObject` for empty structs
  --> tests/ui/invalid_intopy_derive.rs:30:31
   |
30 | struct EmptyTransparentStruct {}
   |                               ^^

error: cannot derive `IntoPyObject` for empty variants
  --> tests/ui/invalid_intopy_derive.rs:35:5
   |
35 |     EmptyTuple(),
   |     ^^^^^^^^^^

error: cannot derive `IntoPyObject` for empty variants
  --> tests/ui/invalid_intopy_derive.rs:42:5
   |
42 |     EmptyStruct {},
   |     ^^^^^^^^^^^

error: transparent structs and variants can only have 1 field
  --> tests/ui/invalid_intopy_derive.rs:48:35
   |
48 | struct TransparentTupTooManyFields(String, String);
   |                                   ^^^^^^^^^^^^^^^^

error: transparent structs and variants can only have 1 field
  --> tests/ui/invalid_intopy_derive.rs:52:39
   |
52 |   struct TransparentStructTooManyFields {
   |  _______________________________________^
53 | |     foo: String,
54 | |     bar: String,
55 | | }
   | |_^

error: transparent structs and variants can only have 1 field
  --> tests/ui/invalid_intopy_derive.rs:60:15
   |
60 |     EmptyTuple(String, String),
   |               ^^^^^^^^^^^^^^^^

error: transparent structs and variants can only have 1 field
  --> tests/ui/invalid_intopy_derive.rs:67:17
   |
67 |       EmptyStruct {
   |  _________________^
68 | |         foo: String,
69 | |         bar: String,
70 | |     },
   | |_____^

error: expected `transparent` or `crate`
  --> tests/ui/invalid_intopy_derive.rs:75:8
   |
75 | #[pyo3(unknown = "should not work")]
   |        ^^^^^^^

error: #[derive(`IntoPyObject`)] is not supported for unions
  --> tests/ui/invalid_intopy_derive.rs:81:1
   |
81 | union Union {
   | ^^^^^

error: cannot derive `IntoPyObject` for empty variants
  --> tests/ui/invalid_intopy_derive.rs:87:5
   |
87 |     Unit,
   |     ^^^^
